File: Develop/app/src/main/java/com/demo/opencv/MainActivity.java
package com.demo.opencv;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.graphics.ImageFormat;
import android.hardware.camera2.CaptureRequest;
import android.media.Image;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.util.Range;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.OptIn;
import androidx.appcompat.app.AppCompatActivity;
import androidx.camera.camera2.interop.Camera2Interop;
import androidx.camera.camera2.interop.ExperimentalCamera2Interop;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.ExperimentalGetImage;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageProxy;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.ViewModelProvider;

import com.demo.opencv.fragments.calibration_display;
import com.demo.opencv.fragments.clinician_mode_display;
import com.demo.opencv.fragments.dev_mode_display;
import com.demo.opencv.fragments.quick_chat_display;
import com.demo.opencv.fragments.social_media_display;
import com.demo.opencv.fragments.text_mode_display;
import com.demo.opencv.other.ClinicalData;
import com.demo.opencv.socialMedia.TwitterAuthenticator;
import com.demo.opencv.vision.ImageFormatUtils;
import com.demo.opencv.vision.Model;
import com.google.common.util.concurrent.ListenableFuture;

import org.json.JSONException;
import org.opencv.core.Mat;

import java.io.IOException;
import java.util.Locale;
import java.util.Objects;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity
        implements ContractInterface.View, calibration_display.OnButtonClickListener,
        text_mode_display.TextModeListener, social_media_display.SocialMediaModeListener,
        quick_chat_display.QuickChatListener

{

    // creating object of Presenter interface in Contract
    ContractInterface.Presenter presenter;
    TwitterAuthenticator twitterAuthenticator;

    public ImageView[] images = new ImageView[4];
    private final int REQUEST_CODE_PERMISSIONS = 1001;
    private final int REQUEST_RECORD_AUDIO_PERMISSION = 200;
    private final String[] REQUIRED_PERMISSIONS = new String[]{"android.permission.CAMERA", "android.permission.RECORD_AUDIO"};
    private final ExecutorService cameraExecutor = Executors.newSingleThreadExecutor();
    private UIViewModel viewModel;
    FragmentManager fragmentManager;
    Handler mainHandler = new Handler(Looper.getMainLooper());
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        UserDataManager userDataManager = (UserDataManager) getApplication();
        userDataManager.getSettings();
        setLocale(userDataManager.getLanguage());
        setContentView(R.layout.activity_main);
        // instantiating object of Presenter Interface
        presenter = new Presenter(this, new Model());

        Log.d("MainActivityGazeLink", "initialized");
        try {
            presenter.initialize(this, getApplication()); // initialize presenter
        } catch (IOException | JSONException e) {
            throw new RuntimeException(e);
        }

        if(allPermissionsGranted()){
            startCamera(); //start camera if permission has been granted by user
        } else{
            ActivityCompat.requestPermissions(this, REQUIRED_PERMISSIONS, REQUEST_CODE_PERMISSIONS);
            ActivityCompat.requestPermissions(this, REQUIRED_PERMISSIONS, REQUEST_RECORD_AUDIO_PERMISSION);
        }
        viewModel = new ViewModelProvider(this).get(UIViewModel.class);

        //twitter authenticator
        twitterAuthenticator = new TwitterAuthenticator(getApplication());
        Intent intent = getIntent();
        Uri uri = intent.getData();
        twitterAuthenticator.handleCallback(uri, getApplication()); // when redirected back, get access token

        // home page set up
        Button devButton, textButton, calibrateButton, clinicianButton, quickChatButton;
        devButton = findViewById(R.id.devModeButton);
        textButton = findViewById(R.id.textModeButton);
        calibrateButton =  findViewById(R.id.calibrateButton);
        clinicianButton = findViewById(R.id.clinicianModeButton);
        quickChatButton = findViewById(R.id.quickChatButton);

        fragmentManager = getSupportFragmentManager();

        openSettings();
        clinicianButton.setOnClickListener(v -> {
            if (!Objects.equals(presenter.getMode(), "Clinician")) {
                setLocale(userDataManager.getLanguage());
                openClinician();
            }
        });
        devButton.setOnClickListener(v -> {
                if (!Objects.equals(presenter.getMode(), "Dev")) {
                    setLocale(userDataManager.getLanguage());
                    openSettings();
                }
            }
        );
        textButton.setOnClickListener(v -> {
            if (!Objects.equals(presenter.getMode(), "Text")) {
                setLocale(userDataManager.getLanguage());
                openTextEntry();
            }
        });
        calibrateButton.setOnClickListener(v -> {
            if (!Objects.equals(presenter.getMode(), "Calibration")) {
                setLocale(userDataManager.getLanguage());
                openCalibration();
            }
        });
        quickChatButton.setOnClickListener(v -> {
            if (!Objects.equals(presenter.getMode(), "QuickChat")) {
                setLocale(userDataManager.getLanguage());
                openQuickChat();
            }
        });
    }

    private void setLocale(String language) {

        String languageCode;
        if (Objects.equals(language, "English")) {
            languageCode = "en-us";
        } else if (Objects.equals(language, "Spanish")) {
            languageCode = "es";
        } else if (Objects.equals(language, "Chinese")) {
            languageCode = "zh";
        } else {
            languageCode = "en-us";
        }
        Log.d("LocaleSettings", "Current Code: " + languageCode);
        Locale locale = new Locale(languageCode);
        Locale.setDefault(locale);
        Resources resources = getResources();
        Configuration configuration = resources.getConfiguration();

        configuration.setLocale(locale);

        resources.updateConfiguration(configuration, resources.getDisplayMetrics());
    }

    @Override
    public void onCalibrationButtonClick() {
        presenter.updateCalibration();
    }

    @Override
    public void updateLiveData(AppLiveData appLiveData) {
        Runnable myRunnable = () -> viewModel.selectItem(appLiveData);
        mainHandler.post(myRunnable);
    }

    @Override
    public void openSocialMedia() {
        presenter.setMode("Social Media");
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainerView, social_media_display.class, null)
                .setReorderingAllowed(true)
                .addToBackStack("SocialMediaFragment") // Name can be null
                .commit();
    }
    @Override
    public void openTextEntry() {
        presenter.setMode("Text");
        Fragment textFragment = text_mode_display.newInstance();
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainerView, textFragment)
                .setReorderingAllowed(true)
                .addToBackStack("text_fragment") // Name can be null
                .commit();
    }
    @Override
    public void openCalibration() {
        presenter.setMode("Calibration");
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainerView, calibration_display.class, null)
                .setReorderingAllowed(true)
                .addToBackStack("calibrate_fragment") // Name can be null
                .commit();
    }
    @Override
    public void openSettings() {
        presenter.setMode("Dev");
        Fragment devFragment = dev_mode_display.newInstance();
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainerView, devFragment)
                .setReorderingAllowed(true)
                .addToBackStack("dev_fragment") // Name can be null
                .commit();
    }

    @Override
    public void openQuickChat() {
        presenter.setMode("QuickChat");
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainerView, quick_chat_display.class, null)
                .setReorderingAllowed(true)
                .addToBackStack("quick_chat_fragment") // Name can be null
                .commit();
    }
    @Override
    public void openClinician() {
        presenter.setMode("Clinician");
        ClinicalData clinicalData = presenter.getClinicalData(); // get clinical data from the presenter
        Fragment calibrationFragment = clinician_mode_display.newInstance(clinicalData);
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainerView, calibrationFragment)
                .setReorderingAllowed(true)
                .addToBackStack("clinician_fragment") // Name can be null
                .commit();
    }



    @Override
    protected void onDestroy() {
        super.onDestroy();
        presenter.onDestroy();
        cameraExecutor.shutdown();
    }

    private void startCamera() {
        final ListenableFuture<ProcessCameraProvider> cameraProviderFuture = ProcessCameraProvider.getInstance(this);
        cameraProviderFuture.addListener(() -> {
            try {
                ProcessCameraProvider cameraProvider = cameraProviderFuture.get();
                bindCamera(cameraProvider);

            } catch (ExecutionException | InterruptedException e) {
                // No errors need to be handled for this Future.
                // This should never be reached.
            }
        }, ContextCompat.getMainExecutor(this));
    }

    @OptIn(markerClass = ExperimentalCamera2Interop.class)
    void bindCamera(@NonNull ProcessCameraProvider cameraProvider) {

        CameraSelector cameraSelector = new CameraSelector.Builder()
                .requireLensFacing(CameraSelector.LENS_FACING_FRONT)
                .build();

        ImageAnalysis.Builder newBuilder = new ImageAnalysis.Builder();

        Camera2Interop.Extender<ImageAnalysis> ext = new Camera2Interop.Extender<>(newBuilder);
        int TARGET_FRAME_RATE = 13;
        ext.setCaptureRequestOption(
                CaptureRequest.CONTROL_AE_TARGET_FPS_RANGE,
                new Range<>(TARGET_FRAME_RATE, TARGET_FRAME_RATE)
        );
        ImageAnalysis imageAnalysis = newBuilder
                .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                .build();
        imageAnalysis.setAnalyzer(cameraExecutor, new ImageAnalysis.Analyzer() {
            @Override
            @OptIn(markerClass = ExperimentalGetImage.class)
            public void analyze(@NonNull ImageProxy image) {

                if (presenter.getPresenterState()) {
                    image.close();
                } else {
                    if (image.getFormat() == ImageFormat.YUV_420_888 && image.getPlanes().length == 3) {
                        //Log.d("MVPView", "Image Format Correct: Frame Loaded");
                        Image newImage = image.getImage();
                        Mat rgbaMat = ImageFormatUtils.rgb(newImage);

                        presenter.onFrame(rgbaMat);
                        assert newImage != null;
                        newImage.close();
                    } else {
                        Log.d("MVPView", "Image format not correct");
                    }
                    image.close();
                }
            }
        });
        cameraProvider.bindToLifecycle(this, cameraSelector, imageAnalysis);
    }

    private boolean allPermissionsGranted(){
        for(String permission : REQUIRED_PERMISSIONS){
            if(ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED){
                return false;
            }
        }
        return true;
    }
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {

        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            if (allPermissionsGranted()) {
                startCamera();
            } else {
                Toast.makeText(this, "Permissions not granted by the user.", Toast.LENGTH_SHORT).show();
                this.finish();
            }
        }
        if (requestCode == REQUEST_RECORD_AUDIO_PERMISSION) {
            Log.d("MVPView", "Record audio permission granted");
        }
    }

    @Override
    public void onTextModeButtonClicked(int input) {
        presenter.onGazeButtonClicked(input);
    }

    @Override
    public void recordButtonClicked() {
        presenter.onRecordButtonClicked();
    }

    @Override
    public void onSocialMediaButtonClicked(int input) { presenter.onGazeButtonClicked(input); }

    @Override
    public void AuthenticationButtonClicked() { // update twitter authentication
        twitterAuthenticator = new TwitterAuthenticator(this);
        twitterAuthenticator.startAuthentication(getApplication());
    }

    @Override
    public void onQuickChatButtonClicked(int input) {
        presenter.onGazeButtonClicked(input);
    }

    @Override
    public void onQuickChatButtonContentChanged(String newText) {

    }
}

File: Develop/app/src/main/java/com/demo/opencv/MainActivity.kt
package com.demo.opencv

// android and java imports

// opencv imports

// other class imports

import android.Manifest
import android.R.attr.angle
import android.R.attr.src
import android.content.pm.PackageManager
import android.graphics.ImageFormat
import android.hardware.camera2.CaptureRequest
import android.media.Image
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.util.Range
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.camera2.interop.Camera2Interop
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageCapture
import androidx.camera.core.ImageProxy
import androidx.camera.core.Preview
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.demo.opencv.databinding.ActivityMainBinding
import org.opencv.android.BaseLoaderCallback
import org.opencv.android.OpenCVLoader
import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.core.Point
import org.opencv.imgproc.Imgproc
import java.nio.ByteBuffer
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

const val TARGET_FRAME_RATE = 12
typealias LumaListener = (luma: Double) -> Unit

open class MainActivity : AppCompatActivity() {
    private lateinit var viewBinding: ActivityMainBinding
    private var imageCapture: ImageCapture? = null
    var eyeDetection = EyeDetection()

    private lateinit var cameraExecutor: ExecutorService

    private val mLoaderCallback: BaseLoaderCallback = object : BaseLoaderCallback(this) {
        override fun onManagerConnected(status: Int) {
            when (status) {
                SUCCESS -> {
                    Log.i("OpenCVDebug", "OpenCV loaded successfully")
                }
                else -> {
                    super.onManagerConnected(status)
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()

        // check and initialize OpenCV
        if (!OpenCVLoader.initDebug()) {
            Log.d("OpenCVDebug", "cannot init debug")
            OpenCVLoader.initAsync(OpenCVLoader.OPENCV_VERSION, this, mLoaderCallback)
        } else {
            Log.d("OpenCVDebug", "success")
            val newContext = this.applicationContext
            eyeDetection.initialize_detector(newContext, this)
        }
        // Request camera permissions
        if (allPermissionsGranted()) {
            startCamera()
        } else {
            ActivityCompat.requestPermissions(
                this, REQUIRED_PERMISSIONS, REQUEST_CODE_PERMISSIONS)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewBinding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(viewBinding.root)
        // Set up the listeners for take photo and video capture buttons
        //viewBinding.imageCaptureButton.setOnClickListener { takePhoto() }
        //viewBinding.videoCaptureButton.setOnClickListener { captureVideo() }

        cameraExecutor = Executors.newSingleThreadExecutor()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>, grantResults:
        IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            if (allPermissionsGranted()) {
                startCamera()
            } else {
                Toast.makeText(this,
                    "Permissions not granted by the user.",
                    Toast.LENGTH_SHORT).show()
                finish()
            }
        }
    }

    inner class ImageAnalyzer(private val listener: LumaListener) : ImageAnalysis.Analyzer {

        private fun ByteBuffer.toByteArray(): ByteArray {
            rewind()    // Rewind the buffer to zero
            val data = ByteArray(remaining())
            get(data)   // Copy the buffer into a byte array
            return data // Return the byte array
        }

        @androidx.annotation.OptIn(androidx.camera.core.ExperimentalGetImage::class)
        override fun analyze(image: ImageProxy) {
            try {
                image.image?.let {
                    // ImageProxy uses an ImageReader under the hood:
                    // https://developer.android.com/reference/androidx/camera/core/ImageProxy.html
                    // That has a default format of YUV_420_888 if not changed that's the default
                    // Android camera format.
                    // https://developer.android.com/reference/android/graphics/ImageFormat.html#YUV_420_888
                    // https://developer.android.com/reference/android/media/ImageReader.html

                    // Sanity check
                    if (it.format == ImageFormat.YUV_420_888
                        && it.planes.size == 3
                    ) {
                        val newImage: Image? = image.getImage()
                        if (newImage != null) {
                            Log.d("ImageProcessing", "imageheight: " +  newImage.height)
                        }
                        var rgbaMat = Yuv420.gray(newImage)
                        if (rgbaMat != null) {
                            Log.d("ImageProcessing", "matimageheight: " +  rgbaMat.height())
                        }
                        eyeDetection.input_detection(rgbaMat)
                    } else {
                        // Manage other image formats
                        // TODO - https://developer.android.com/reference/android/media/Image.html
                    }
                }
            } catch (ise: IllegalStateException) {
                ise.printStackTrace()
            }

            image.close()
        }
    }

    @androidx.annotation.OptIn(androidx.camera.camera2.interop.ExperimentalCamera2Interop::class)
    private fun startCamera() {
        val cameraProviderFuture = ProcessCameraProvider.getInstance(this)
        cameraProviderFuture.addListener({
            // Used to bind the lifecycle of cameras to the lifecycle owner
            val cameraProvider: ProcessCameraProvider = cameraProviderFuture.get()

            // Preview
            val preview = Preview.Builder()
                .build()
                .also {
                    it.setSurfaceProvider(viewBinding.viewFinder.surfaceProvider)
                }

            imageCapture = ImageCapture.Builder()
                .build()

            val builder = ImageAnalysis.Builder()
            val ext: Camera2Interop.Extender<*> = Camera2Interop.Extender(builder)
            ext.setCaptureRequestOption(
                CaptureRequest.CONTROL_AE_TARGET_FPS_RANGE,
                Range<Int>(TARGET_FRAME_RATE,TARGET_FRAME_RATE)
            )
            val imageAnalyzer = builder // set image analysis settings here (resolution, color, frame rate, etc.)
                .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                .build()
                .also {
                    it.setAnalyzer(cameraExecutor, ImageAnalyzer { luma ->
                       // Log.d("Luminosity!", "Average luminosity: $luma")
                    })
                }

            // Select front camera as a default
            val cameraSelector = CameraSelector.DEFAULT_FRONT_CAMERA

            try {
                // Unbind use cases before rebinding
                cameraProvider.unbindAll()

                // Bind use cases to camera
                cameraProvider.bindToLifecycle(
                    this, cameraSelector, preview, imageCapture, imageAnalyzer)

            } catch(exc: Exception) {
                Log.e(TAG, "Use case binding failed", exc)
            }

        }, ContextCompat.getMainExecutor(this))
    }

    private fun allPermissionsGranted() = REQUIRED_PERMISSIONS.all {
        ContextCompat.checkSelfPermission(
            baseContext, it) == PackageManager.PERMISSION_GRANTED
    }

    override fun onDestroy() {
        super.onDestroy()
        cameraExecutor.shutdown()
    }

    companion object {
        private const val TAG = "CameraXApp"
        private const val FILENAME_FORMAT = "yyyy-MM-dd-HH-mm-ss-SSS"
        private const val REQUEST_CODE_PERMISSIONS = 10
        private val REQUIRED_PERMISSIONS =
            mutableListOf (
                Manifest.permission.CAMERA,
                Manifest.permission.RECORD_AUDIO
            ).apply {
                if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) {
                    add(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                }
            }.toTypedArray()
    }
}